### PR TITLE
Adding 3D/2D renderer system

### PR DESCRIPTION
**Pull request by Octopus773**
## Proposed changes

For the fact that tests aren't building
https://github.com/AnonymusRaccoon/Bomberman/pull/64#issuecomment-848597247

Adding
- 2D Renderer System
- 2D Component
- 3D Renderer System
- 3D Component
- A big load of assets

Refactoring
- All RAY Drawables .cpp files have been modified (Namesapce fix)

## Information
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking changes

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

## Known Bugs
- RenderScreenSystem doesn't compile with camera2d (this camera lack some member functions)

### CODE REVIEW

**Comment by Octopus773** [comment]
Test build is failing, waiting for pr https://github.com/AnonymusRaccoon/Bomberman/pull/63 to be merged for fixing this issue

File: lib/Ray/sources/Window.hpp [hunk 1]
```diff
@@ -111,9 +112,13 @@ namespace RAY {
 			void setTitle(const std::string &title);
 
 
-			//! @brief draw rectangle
+			//! @brief draw drawable
 			//! @param drawable The drawable to render on screen
-			void draw(Drawables::IDrawable &drawable) override;
```
Comment (AnonymusRaccoon) [inline lib/Ray/sources/Window.hpp:116]: Why did you change this? You should take the lowest type that fit your needs as parameters, here a IDrawable is enough since they all have a drawOn(Window) method. There is no need to specify a specific method for 2D and 3D.

File: sources/System/Renderer/Renderer3DSystem.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+		//! @brief Update the corresponding component of the given entity
+		//! @param entity The entity to update.
+		//! @param dtime The delta time.
+		void onUpdate(WAL::Entity &entity, std::chrono::nanoseconds dtime) override
+		{
+			auto &comp = entity.getComponent<Drawable3DComponent<T>>();
+			auto &pos = entity.getComponent<PositionComponent>();
+
+			comp.member.setPosition({pos.getX(), pos.getY(), pos.getZ()});
```
Comment (AnonymusRaccoon) [inline sources/System/Renderer/Renderer3DSystem.hpp:38]: You should implement a conversion operator in the Vector3 class to simply pass the vector to this method.

File: tests/EntityTests.cpp [hunk 3]
```diff
@@ -6,6 +6,7 @@
 #include "Component/Position/PositionComponent.hpp"
 #include <catch2/catch.hpp>
 
+using namespace BBM;
```
Comment (AnonymusRaccoon) [inline tests/EntityTests.cpp:9]: You don't need to use the same namespace twice.

**Review by AnonymusRaccoon — COMMENTED** [review]
Changing namespace or public API is called a breaking change, you should update your PR description accordingly.

File: tests/EntityTests.cpp (hunk 3 above)
Comment (Octopus773) [reply tests/EntityTests.cpp:9]: It was a merge isue

**Review by Octopus773 — COMMENTED** [review]
test review from clion

**Review by Octopus773 — COMMENTED** [review]
pppp

File: lib/Ray/sources/Drawables/IDrawable.hpp [hunk 4]
```diff
@@ -24,7 +22,7 @@ namespace RAY
 				virtual void drawOn(RAY::Window &) = 0;
 			protected:
 			private:
-		};
+		};p
```
Comment (AnonymusRaccoon) [inline lib/Ray/sources/Drawables/IDrawable.hpp:25]: Are you sure about this 'p'? x)

Comment (Octopus773) [reply lib/Ray/sources/Drawables/IDrawable.hpp]: i just saw it xD

File: sources/System/Renderer/RenderScreenSystem.hpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+#pragma once
+
+#include "System/System.hpp"
+#include "Camera/Camera2D.hpp"
+#include "Window.hpp"
+
+namespace BBM
+{
+	template <class  T>
```
Comment (AnonymusRaccoon) [inline sources/System/Renderer/RenderScreenSystem.hpp:13]: Why is this templated?

File: sources/System/Movable/MovableSystem.hpp [hunk 6]
```diff
@@ -6,11 +6,12 @@
 #pragma once
 
 #include "System/System.hpp"
+#include "Entity/Entity.hpp"
 
 namespace BBM
 {
 	//! @brief A system to handle movable entities. This system update velocity based on accelerations and positions based on velocity.
-	class MovableSystem : public WAL::System
+class MovableSystem : public WAL::System
```
Comment (TrueBabyChaise) [inline sources/System/Movable/MovableSystem.hpp:14]: Wrong indentation

File: sources/System/Renderer/RenderScreenSystem.hpp [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	class RenderScreenSystem : public WAL::System
+	{
+		//! @brief The window to render on
+		RAY::Window &_window;
+		//! @brief The camera
+		T &_camera;
+	public:
+		//! @brief ctor
+		explicit RenderScreenSystem(RAY::Window &window, T &camera)
+			: WAL::System({typeid(RenderScreenSystem)}),
```
[the remark this answers is not in the document]
Comment (Octopus773) [reply sources/System/Renderer/RenderScreenSystem.hpp:23]: ~~indeed i can~~
i set an empty list

File: sources/Util/Utils.cpp [hunk 8]
```diff
@@ -7,8 +7,12 @@
 
 #include "Vector/Vector3.hpp"
 #include "Models/Vector3.hpp"
+#include "Utils.hpp"
 
-RAY::Vector3 toRAY(const WAL::Vector3f &wal)
+namespace BBM
 {
-	return RAY::Vector3(wal.x, wal.y, wal.y);
+	RAY::Vector3 Utils::toRAY(const BBM::Vector3f &wal)
```
Comment (AnonymusRaccoon) [inline sources/Util/Utils.cpp:14]: This function could be a conversion operator inside the Vector3 class (at first it was a function because the Vector3 was inside the WAL library but since the Vector has been moved, this function has no reason to exist outside of the class).

**Review by TrueBabyChaise — APPROVED** [review]

**Review by EternalRat — APPROVED** [review]

File: sources/Util/Utils.hpp [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+#include "Vector/Vector3.hpp"
+#include "Models/Vector3.hpp"
+
+namespace BBM
+{
+	struct Utils
+	{
+		//! @brief Convert BBM Vector3f to RAY Vector3
+		static RAY::Vector3 toRAY(const BBM::Vector3f &wal);
+p
```
Comment (AnonymusRaccoon) [inline sources/Util/Utils.hpp]: Nice p

File: sources/Util/Utils.hpp [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+		//! @brief default ctor
+		Utils() = default;
```
Comment (AnonymusRaccoon) [inline sources/Util/Utils.hpp:18]: Don't put constructors in a static class (a class where every member is static)

File: sources/System/Renderer/RenderScreenSystem.hpp (hunk 5 above)
Comment (Octopus773) [reply sources/System/Renderer/RenderScreenSystem.hpp:13]: To allow multiple types of cameras
here 3D and 2D (2D doesn't support it due to missing functions of the wrapper)

File: sources/Util/Utils.hpp (hunk 10 above)
Comment (AnonymusRaccoon) [reply sources/Util/Utils.hpp:18]: This class can be deleted anyway

File: sources/Util/Utils.hpp (hunk 9 above)
Comment (Octopus773) [reply sources/Util/Utils.hpp]: fixed ;)

File: sources/Util/Utils.hpp (hunk 10 above)
Comment (Octopus773) [reply sources/Util/Utils.hpp:18]: nah i follow cooplean form

Comment (Octopus773) [reply sources/Util/Utils.hpp:18]: even with useless stuff

File: sources/System/Renderer/RenderScreenSystem.hpp (hunk 5 above)
Comment (AnonymusRaccoon) [reply sources/System/Renderer/RenderScreenSystem.hpp:13]: You can't use both systems at once so templating this system is missleading. You can put a boolean 2D or 3D if you want both or something like that but templating it seems like a bad idea.

File: sources/Util/Utils.cpp (hunk 8 above)
Comment (Octopus773) [reply sources/Util/Utils.cpp:14]: As you please -> https://github.com/AnonymusRaccoon/Bomberman/pull/64/commits/e70dbd5b3b428d516d2e190c2e117f6eed62b01b

File: sources/Models/Vector3.cpp [hunk 11]
```diff
@@ -0,0 +1,13 @@
+//
+// Created by cbihan on 26/05/2021.
+//
+
+#include "Vector3.hpp"
+
+namespace BBM
+{
+	RAY::Vector3 operator*(const Vector3f &v)
```
Comment (GitBluub) [inline sources/Models/Vector3.cpp:9]: operator* should take void if you want to use it as *vector3, this one is for arithmetic operator

Comment (AnonymusRaccoon) [reply sources/Models/Vector3.cpp:9]: Yes, I was talking about a conversion operator. See [this](https://en.cppreference.com/w/cpp/language/cast_operator).

File: sources/Util/Utils.hpp [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+#include "Vector/Vector3.hpp"
+#include "Models/Vector3.hpp"
+
+namespace BBM
+{
+	struct Utils
+	{
+		//! @brief Convert BBM Vector3f to RAY Vector3
+		static RAY::Vector3 toRAY(const BBM::Vector3f &wal);
```
Comment (GitBluub) [inline sources/Util/Utils.hpp:15]: Don't need this anymore if we use the *operator

Comment (Octopus773) [reply sources/Util/Utils.hpp]: fixed -> https://github.com/AnonymusRaccoon/Bomberman/pull/64/commits/6bd7a46ea978636174595209c16c3a0b4e942f56

File: sources/Models/Vector3.cpp (hunk 11 above)
Comment (Octopus773) [reply sources/Models/Vector3.cpp:9]: fixed -> https://github.com/AnonymusRaccoon/Bomberman/pull/64/commits/07e2d5bc35aadd0db4cf31ef5a71ac19f06a42c9

**Review by Arthi-chaud — APPROVED** [review]
